File: Yu/Sources/Test-Driven-Development-By-Example/Sum.swift
public class Sum: Expression {
  public init(augend: Expression, addend: Expression) {
    self.augend = augend
    self.addend = addend
  }

  public var augend: Expression
  public var addend: Expression

  public func reduce(bank: Bank, _ to: String) -> Money {
    let amount: Int = augend.reduce(bank: bank, to).amount + addend.reduce(bank: bank, to).amount
    return Money(amount, currency: to)
  }

  public func plus(_ addend: Expression) -> Expression {
    Sum(augend: self, addend: addend)
  }
  public func times(_ multiplier: Int) -> Expression {
    Sum(augend: augend.times(multiplier), addend: addend.times(multiplier))
  }
}

File: clark/src/Sum/Sum.js
class Sum {
  constructor(augend, addend) {
    this.augend = augend;
    this.addend = addend;
  }

  reduce(bank, to) {
    const Money = require('../Money/Money.js');
    const amount = this.augend.reduce(bank, to).amount
      + this.addend.reduce(bank, to).amount;
    return new Money(amount, to);
  }

  plus(addend) {
    return new Sum(this, addend);
  }

  times(multiplier) {
    return new Sum(this.augend.times(multiplier), this.addend.times(multiplier));
  }
}

module.exports = Sum;
